materialize-postgres/sqlgen.go

```diff
@@ -10,62 +10,43 @@ import (
-					// Postgres doesn't allow fields with null bytes, so they must be stripped out if
-					// present.
-					return strings.ReplaceAll(in, "\u0000", ""), nil
-				})),
+	mapper := sql.NewDDLMapper(
+		sql.FlatTypeMappings{
+			sql.INTEGER: sql.MapSignedInt64(
+				sql.MapStatic("BIGINT", sql.AlsoCompatibleWith("integer"), sql.UsingConverter(sql.CheckedInt64)),
+				sql.MapStatic("NUMERIC"),
 			),
-			WithFormat: map[string]sql.TypeMapper{
-				"integer": sql.PrimaryKeyMapper{
-					PrimaryKey: sql.NewStaticMapper("TEXT"),
-					Delegate:   sql.NewStaticMapper("NUMERIC"),
+			sql.NUMBER:         sql.MapStatic("DOUBLE PRECISION"),
+			sql.BOOLEAN:        sql.MapStatic("BOOLEAN"),
+			sql.OBJECT:         sql.MapStatic("JSON"),
+			sql.ARRAY:          sql.MapStatic("JSON"),
+			sql.BINARY:         sql.MapStatic("TEXT", sql.AlsoCompatibleWith("character varying")),
+			sql.MULTIPLE:       sql.MapStatic("JSON", sql.UsingConverter(sql.ToJsonBytes)),
+			sql.STRING_INTEGER: sql.MapStatic("NUMERIC"),
+			sql.STRING_NUMBER:  sql.MapStatic("DECIMAL", sql.AlsoCompatibleWith("numeric")),
+			sql.STRING: sql.MapString(sql.StringMappings{
+				Fallback: sql.MapStatic(
+					"TEXT",
+					sql.AlsoCompatibleWith("character varying"),
+					sql.UsingConverter(sql.StringCastConverter(func(in string) (any, error) {
+						// Postgres doesn't allow fields with null bytes, so they must be stripped out if
+						// present.
+						return strings.ReplaceAll(in, "\u0000", ""), nil
+					})),
+				),
+				WithFormat: map[string]sql.MapProjectionFn{
+					"date":      sql.MapStatic("DATE", sql.UsingConverter(sql.ClampDate)),
+					"date-time": sql.MapStatic("TIMESTAMPTZ", sql.AlsoCompatibleWith("timestamp with time zone"), sql.UsingConverter(sql.ClampDatetime)),
+					"duration":  sql.MapStatic("INTERVAL"),
+					"ipv4":      sql.MapStatic("CIDR"),
+					"ipv6":      sql.MapStatic("CIDR"),
+					"macaddr":   sql.MapStatic("MACADDR"),
+					"macaddr8":  sql.MapStatic("MACADDR8"),
+					"time":      sql.MapStatic("TIME", sql.AlsoCompatibleWith("time without time zone")),
 				},
-				"number": sql.PrimaryKeyMapper{
-					PrimaryKey: sql.NewStaticMapper("TEXT"),
-					Delegate:   sql.NewStaticMapper("DECIMAL"),
-				},
-				"date":      sql.NewStaticMapper("DATE", sql.WithElementConverter(sql.ClampDate())),
-				"date-time": sql.NewStaticMapper("TIMESTAMPTZ", sql.WithElementConverter(sql.ClampDatetime())),
-				"duration":  sql.NewStaticMapper("INTERVAL"),
-				"ipv4":      sql.NewStaticMapper("CIDR"),
-				"ipv6":      sql.NewStaticMapper("CIDR"),
-				"macaddr":   sql.NewStaticMapper("MACADDR"),
-				"macaddr8":  sql.NewStaticMapper("MACADDR8"),
-				"time":      sql.NewStaticMapper("TIME"),
-			},
+			}),
 		},
-	}
-
-	mapper = sql.NullableMapper{
-		NotNullText: "NOT NULL",
-		Delegate:    mapper,
-	}
-
-	columnValidator := sql.NewColumnValidator(
-		sql.ColValidation{Types: []string{"bigint", "integer"}, Validate: sql.IntegerCompatible},
```

Comment:
IIRC there was a single case of a pre-existing table where a column type of `integer` needed to be allowed. If this task is still around it an use a custom DDL field config.

Reply:
Note to self: Our stats materialization actually will require this. That might have been the only case where this is needed.